webui: introduce subscription backup unit report

fbergkemper opened the pull request:
Introduces an additional view to the director module where a backup unit report can be retrieved and downloaded in JSON format.

### Thank you for contributing to the Bareos Project!

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)


### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)

##### General
- [x] PR name is meaningful
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
- [x] Check backport line
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Separate commit for CHANGELOG.md ("update CHANGELOG.md"). The PR number is correct.

##### Source code quality

- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [ ] Required documentation changes are present and part of the PR
- [x] `bareos-check-sources --since-merge` does not report any problems

##### Tests

- [ ] Decision taken that a test is required (if not, then remove this paragraph)
- [ ] The choice of the type of test (unit test or systemtest) is reasonable
- [ ] Testname matches exactly what is being tested
- [ ] On a fail, output of the test leads quickly to the origin of the fault

joergsteffens commented on webui/install/bareos/bareos-dir.d/profile/webui-admin.conf:
```diff
@@ -3,7 +3,7 @@
 Profile {
   Name = "webui-admin"
-  CommandACL = !.bvfs_clear_cache, !.exit, !.sql, !configure, !create, !delete, !purge, !prune, !sqlquery, !umount, !unmount, *all*
+  CommandACL = !.bvfs_clear_cache, !.exit, !.sql, !create, !delete, !purge, !prune, !sqlquery, !umount, !unmount, *all*
```
What for the `configure` command is needed? With `configure` allowed, all other ACL restrictions are useless.

fbergkemper replied:
Maybe @arogge can give us some insides why this is the case, I did not expect the behavior as well.

joergsteffens replied:
Well, the case is obvious. For `status subscription` you need to know all resources. Internally, a check for the configure command ACL is often (mis-)used to detect, if you have "unlimited" access.
I'm not sure, what resources `status subscription` really check. I guess, this check could get replaced by checks for access to ALL clients and ALL jobs. Maybe some more. So I guess, this is what is required. Next question is, who will implement this and should this get implemented inside this PR or a separate one.

joergsteffens replied:
... I guess, `status subscription` will also work with limited access on jobs and clients. However, than information will than not cover the full system.

fbergkemper replied:
For the record the error message you would receive.
`
string(208) "{"jsonrpc":"2.0","id":null,"error":{"code":1,"message":"failed","data":{"result":{},"messages":{"error":["status subscription: is an invalid command or needs access right to the \"configure\" command.\n"]}}}}"
`

fbergkemper replied:
https://github.com/bareos/bareos/pull/1280/commits/4b627ffc2e6beb9877563da722eb85197e9a7445 should eliminate our concerns giving access to the configure command, as it would not be required anymore.

fbergkemper replied:
Rebased once again, https://github.com/bareos/bareos/commit/441032cf7ea1355fa35724d80ad16b28863a1e19 should be a better solution.

fbergkemper replied:
outdated

joergsteffens reviewed (changes requested):
As discussed, some improvements. Please you this as a bases for further changes.

fbergkemper commented:
Applied requested changes and cleaned up the PR. Changed the wording in some places as well. Please have another review.

fbergkemper commented:
Changes pushed including some further suggestions. Please review, thank you.

joergsteffens commented on webui/module/Application/config/commands.csv:
```diff
@@ -1,83 +1,83 @@
-.bvfs_restore,--,optional,--,--,--,--,--,--,required,--,--,--
-.bvfs_cleanup,--,optional,--,--,--,--,--,--,--,--,--,--
-.bvfs_clear_cache,--,optional,--,--,--,--,--,--,--,--,--,--
+Command,Client,Console,Dashboard,Director,Fileset,Job,Media,Pool,Restore,Schedule,Storage,Analytics,Subscription
```
Is the new Subscription column still required/useful?

fbergkemper replied:
This is outdate, has already been removed.

joergsteffens commented on webui/module/Director/src/Director/Controller/DirectorController.php:
```diff
@@ -90,6 +90,108 @@ public function indexAction()
+    ));
+  }
+
+  public function backupunitreportAction()
```
The functions subscriptionAction and backupunitreportAction are nearly identical. Can you please modify it to reuse the common code?

fbergkemper replied:
Controller actions are already reduced to a limit to what needs to be there to deliver a new view model. Maybe we could argue on a refactor of the plugins called to get rid of conditions, but that should be done in a separate PR, because it affects all controllers and modules. It would be a bigger change which should not be part of this PR. If we really would like to touch this, we should create a story for that to optimize controller actions and plugin code, I think.

joergsteffens commented on webui/module/Director/src/Director/Model/DirectorModel.php:
```diff
@@ -86,6 +86,38 @@ public function getDirectorStatus(&$bsock=null)
+      return $result;
+   }
+
+   public function getStatusSubscriptionAll(&$bsock=null)
```
Is it possible to unify the two functions? Like:
`getStatusSubscription(&$bsock=null, option="", json=0)`
with  `option = "" | "all" | detail`
If not, the function names should indicate, what the function does. Both function have nearly the same name, but one is returning text, the other a json dictionary.

fbergkemper replied:
Unified, see https://github.com/bareos/bareos/pull/1280/commits/936698c227172bcd303b900e91dd426978b33cb7 .

fbergkemper replied:
Updated: https://github.com/bareos/bareos/pull/1280/commits/bd89100dfd7efe20c322ba079f6c7f47d69fb3b7

joergsteffens reviewed (changes requested):
As discussed, I updated the text on the subscription page, tested it again and fixed some small thinks.
What still needs improvement is the handling of insufficient permissions:
Insufficient permissions are not clearly indicated. Instead, the command output is just displayed:
`status subscription: needs access to all client, job and fileset resources.`
Expected result: warning banner: insufficient permissions, required .... and the download button should also be invisible.

joergsteffens reviewed (approved):
I approve the changes. Please squash the commits before merging.

fbergkemper commented:
Removed core changes from this branch that have already been merged into master by another PR. See https://github.com/fbergkemper/bareos/commit/4cef162b6ec0e669ff3acf4a23281b65a164a79d ( https://github.com/bareos/bareos/pull/1312 ). Squashed the commits and pushed a CHANGELOG.md update.
